### PR TITLE
Optimize fits.Header parsing

### PR DESCRIPTION
**Pull request by saimn**
Related to #5593, this implements one of the idea mentioned there: adding a `fromcards` class method to speed up the creation of a Header instance from a list of cards. 
The other main change is some reorganization of keyword parsing (`_parse_keyword`), improving the most common case.
And probably a small bugfix with the return value from `_check_if_rvkc_image` which was not used. This was probably harmless, I guess the check was done again later.

The gain is significant (~12-13%) for files with many HDUs, below just computing the number of HDUs for a file with >300 extensions (the second command is using 100 extensions, the third only 10). 

Before:
```
~/lib/astropy/fits-header
❯ python -m timeit -n 1 -r 5 -s "from astropy.io import fits" "print(len(fits.open('./scipost_white.fits')))"
336
336
336
336
336
1 loop, best of 5: 6.16 sec per loop

~/lib/astropy/fits-header 34s
❯ python -m timeit -n 1 -r 5 -s "from astropy.io import fits" "print(len(fits.open('./scipost_white.fits')[:100]))"
100
100
100
100
100
1 loop, best of 5: 1.85 sec per loop

~/lib/astropy/fits-header
❯ python -m timeit -n 5 -r 5 -s "from astropy.io import fits" "print(len(fits.open('./scipost_white.fits')[:10]))"
10
...
10
5 loops, best of 5: 167 msec per loop
```
After:
```
~/lib/astropy/fits-header
❯ python -m timeit -n 1 -r 5 -s "from astropy.io import fits" "print(len(fits.open('./scipost_white.fits')))"
336
336
336
336
336
1 loop, best of 5: 5.38 sec per loop

~/lib/astropy/fits-header 28s
❯ python -m timeit -n 1 -r 5 -s "from astropy.io import fits" "print(len(fits.open('./scipost_white.fits')[:100]))"
100
100
100
100
100
1 loop, best of 5: 1.64 sec per loop

~/lib/astropy/fits-header
❯ python -m timeit -n 5 -r 5 -s "from astropy.io import fits" "print(len(fits.open('./scipost_white.fits')[:10]))"
10
...
10
5 loops, best of 5: 152 msec per loop
```
I'm also working on other ideas from  #5593, but this will take more time, so let's go with this for now!

### CODE REVIEW

File: astropy/io/fits/card.py [hunk 1]
```diff
@@ -24,6 +24,7 @@
 KEYWORD_LENGTH = 8  # The max length for FITS-standard keywords
 
 VALUE_INDICATOR = '= '  # The standard FITS value indicator
+VALUE_INDICATOR_LEN = 2
```
Comment (MSeifert04) [inline astropy/io/fits/card.py]: Wouldn't it be more semantically correct to replace the `2` with `len(VALUE_INDICATOR)`?

File: astropy/io/fits/header.py [hunk 2]
```diff
@@ -93,7 +93,9 @@ def __init__(self, cards=[], copy=False):
 
             .. versionadded:: 1.3
         """
-        self.clear()
+        self._cards = []
```
Comment (MSeifert04) [inline astropy/io/fits/header.py]: Hm, why this change? Isn't this just duplicating code?

File: astropy/io/fits/header.py [hunk 3]
```diff
@@ -449,6 +451,19 @@ def block_iter(nbytes):
             if close_file:
                 fileobj.close()
 
+    @classmethod
+    def _fromcards(cls, cards):
+        header = cls()
+        for idx, card in enumerate(cards):
+            header._cards.append(card)
+            keyword = Card.normalize_keyword(card.keyword)
```
Comment (MSeifert04) [inline astropy/io/fits/header.py]: This seems to be copied from the `append` method, couldn't this be factored into a separate method?

File: astropy/io/fits/header.py [hunk 4]
```diff
@@ -938,13 +953,14 @@ def keys(self):
         instance has the same behavior.
         """
 
-        return self.__iter__()
```
Comment (MSeifert04) [inline astropy/io/fits/header.py]: Why was this changed? It seems unlikely this will affect performance, does it?

**Comment by MSeifert04** [comment]
Could you share the fits or the code to produce one that's like it?

File: astropy/io/fits/card.py (hunk 1 above)
Comment (saimn) [reply astropy/io/fits/card.py]: Yes, but I hard coded the value to avoid a tiny overhead at import time.

File: astropy/io/fits/header.py (hunk 4 above)
Comment (saimn) [reply astropy/io/fits/header.py]: About this and `.clear` above, I included a commit to change a few non-obvious lines of code:
https://github.com/astropy/astropy/pull/8428/commits/04f2c36b109125554b6bda4e9558d91a0fb2aef2
I think that these changes help a lot in the understanding of the flow in this complex class. For instance in `__init__` I prefer to duplicate three lines instead of having to search each time where some variable is defined.

File: astropy/io/fits/header.py (hunk 3 above)
Comment (saimn) [reply astropy/io/fits/header.py]: If you can find a good name yes ;)

**Comment by saimn** [comment]
I cannot share this specific file, I could build a similar one but this should noticeable for any file with many keywords (and not specifically for files with many HDUs - I updated the issue title -, this will be a following PR). So it would be useful if you can test on other files ;)

Here is the profiler result, a lot a time spent in `Header.append` (hence the new `_fromcards`) and in keyword parsing:
capture d ecran de 2019-02-17 21-49-28

File: astropy/io/fits/card.py (hunk 1 above)
Comment (MSeifert04) [reply astropy/io/fits/card.py]: Is it worth it?

File: astropy/io/fits/header.py (hunk 3 above)
Comment (MSeifert04) [reply astropy/io/fits/header.py]: `_add_card`?

File: astropy/io/fits/header.py (hunk 4 above)
Comment (MSeifert04) [reply astropy/io/fits/header.py]: But it makes it harder to understand the semantics. Searching for definitions is an `IDE` thing while understanding the semantics and invariants is a Developer thing. And fits is complicated enough we shouldn't make it harder.

This is just me thinking. If that makes it easier for you to maintain the module go ahead (you're a lot more active than me so it's your judgement call). :)

File: astropy/io/fits/card.py (hunk 1 above)
Comment (saimn) [reply astropy/io/fits/card.py]: The overhead is probably negligible, I can change if you prefer. But the value is hardcoded and defined just above so using `len` seems overkill.

File: astropy/io/fits/header.py (hunk 3 above)
Comment (saimn) [reply astropy/io/fits/header.py]: The method could not contain the `_cards.append` since `.append` can also insert the card at any position ... so it would just take care of the indices.

File: astropy/io/fits/header.py (hunk 4 above)
Comment (saimn) [reply astropy/io/fits/header.py]: I don't think that the change makes thing harder ;), if you look at the `items`, `keys` and `values` definition it is also much more consistent and easier to understand now.

Comment (MSeifert04) [reply astropy/io/fits/header.py]: Okay sorry, maybe this was the worst comment for my argument. 😅  

I don't mind the changes to `keys` and `values` (much).

File: astropy/io/fits/header.py (hunk 2 above)
Comment (MSeifert04) [reply astropy/io/fits/header.py]: I don't think this is a good change. It's identical to `clear` and the method `clear` seems much more obvious to me (even without looking at the method) than these 3 re-assignments.

I'd prefer the previous call `self.clear()` here.

File: astropy/io/fits/header.py (hunk 3 above)
Comment (MSeifert04) [reply astropy/io/fits/header.py]: Could you elaborate? I don't understand the previous comment.

**Comment by MSeifert04** [comment]
Thanks for the profiling!

May seem like a stupid idea but could you put the non-performance changes into a different PR, because I really like the performance improvements and I don't want the PR to suffer from my nitpicking comments about the non-performance-related changes? Or is that too much trouble?

File: astropy/io/fits/card.py (hunk 1 above)
Comment (MSeifert04) [reply astropy/io/fits/card.py]: Maybe overkill, but I only commented because it took me a few seconds to figure out what this value is (probably more time than saved by hardcoding it 😄).

Comment (MSeifert04) [reply astropy/io/fits/card.py]: Thanks for changing it 👍

**Comment by saimn** [comment]
Ok, I reverted the change for the `clear` call, I still find it weird to have to call a `clear` method as the first line of `__init__` but this is not a big deal. I have another branch with a dozen commits based on this one so I would prefer not having to split things now, and the cosmetic changes are rather limited.

**Comment by pllim** [comment]
CircleCI failure is #8431 and unrelated. Since there are 2 approvals, I am merging this. Thanks!

**Comment by pllim** [comment]
By the way, do we need a benchmark for this over at https://github.com/astropy/astropy-benchmarks ?

**Comment by saimn** [comment]
Thanks for the reviews.

@mhvk - I don't like it too, but the overhead of a function call is not negligible when it is done for thousands of keywords. Would be great if Python could optimize this itself.

@pllim - Yes, it would be great. I will see of I can add some benchmarks.
